### PR TITLE
Add support for generating Schemas with Java8 Optionals, fixes #1144

### LINKED ISSUE #1144: Provide the option for using Java8 Optionals in generated schema code
As a developer of a product using AtlasDB, I wish to be able to have my schema use Java8's Optionals rather than Guava's Optionals.

Caveat: this should probably be controlled by a flag, so as not to break all AtlasDB clients.

By request of @uschi2000

### PR DESCRIPTION
**Pull request by uschi2000**
- Before this change, only Guava Optional is supported
    - Introduces a breaking change to Schema.class constructor
    - Changes the sort order of generated {@link Foo} code (not a functional break)

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1162)

### CODE REVIEW

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/SweepSchema.java [hunk 1]
```diff
@@ -39,7 +39,8 @@ public Schema get() {
     private static Schema generateSchema() {
         Schema schema = new Schema("Sweep",
                 SweepSchema.class.getPackage().getName() + ".generated",
-                NAMESPACE);
+                NAMESPACE,
+                false);
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/SweepSchema.java]: Might consider enum instead indicating the generation target.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java [hunk 2]
```diff
@@ -1296,89 +1319,96 @@ public boolean apply(IndexMetadata index) {
...
-    };
+    private List<String> getImports() {
+        List<String> imports = Lists.newArrayList();
+        if (useJava8Optionals) {
+            imports.add("java.util.Optional");
+        } else {
+            imports.add(Optional.class.getCanonicalName());
+        }
+        imports.addAll(
+        Arrays.asList(Set.class.getCanonicalName(),
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java]: Might be nicer to build List<Class<?>> and transform to canonical name to reduce chance of errors

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/Schema.java [hunk 3]
```diff
@@ -264,7 +270,7 @@ public void renderTables(File srcDir) throws IOException {
         Preconditions.checkNotNull(name, "schema name not set");
         Preconditions.checkNotNull(packageName, "package name not set");
 
-        TableRenderer tableRenderer = new TableRenderer(packageName, namespace);
+        TableRenderer tableRenderer = new TableRenderer(packageName, namespace, useJava8Optionals);
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/Schema.java]: Yeah, definitely prefer enum over Boolean flag

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java (hunk 2 above)
Comment (gsheasby) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java]: Did we need to switch from `ClassName.class` to calling `.getCanonicalName()` everywhere?

File: atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+public class SchemaTest {
+
+    @Rule
+    public TemporaryFolder testFolder = new TemporaryFolder();
+
+    private static TableReference TABLE_REF = TableReference.createWithEmptyNamespace("TestTable");
+
+    @Test
+    public void testRenderRespectsOptionalsConfiguration() throws IOException {
```
Comment (gsheasby) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java]: This should be split into two separate tests. (Tests should also follow the One Thing Rule).

File: atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/render/TableRendererTest.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import com.palantir.atlasdb.table.description.TableDefinition;
+import com.palantir.atlasdb.table.description.ValueType;
+
+public class TableRendererTest {
+
+    private static TableReference TABLE_REF = TableReference.createWithEmptyNamespace("TestTable");
+    private static SortedSet<IndexMetadata> NO_INDICES = new TreeSet<>();
+
+    @Test
+    public void testRenderRespectsOptionalsConfiguration() {
```
Comment (gsheasby) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/render/TableRendererTest.java]: Again, this is actually two tests.

File: docs/source/release_notes/release-notes.rst [hunk 6]
```diff
@@ -42,6 +42,10 @@ develop
          - Change
 
     *    - |improved|
+         - Add support for generating schemas with Java8 Optionals instead of Guava Optionals. Note that this change
```
Comment (gsheasby) [inline docs/source/release_notes/release-notes.rst]: currently, this is a breaking change, as all products will have to change all of their schemas (even historical ones). Fixing the constructors will fix this (and also mean not having to do so many changes to constructors).

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/Schema.java [hunk 7]
```diff
@@ -67,6 +67,11 @@
     private final String name;
     private final String packageName;
     private final Namespace namespace;
+    /**
+     * Indicates whether the generated code uses the Java8 Optional class (if true) or the Guava Optional class (if
+     * false).
+     */
+    private final boolean useJava8Optionals;
```
Comment (uschi2000) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/Schema.java]: Done (in this PR)

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java [hunk 8]
```diff
@@ -1043,22 +1050,38 @@ private void renderGetAllRowsUnordered() {
...
+                    renderOptionalReturnValue("rowResult");
                 } line("}");
             } line("}");
         }
 
+        private String getOptionalTypeForClass() {
+            return "Optional<" + RowResult + ">";
+        }
+
+        private void renderOptionalReturnValue(@Nullable String value) {
```
Comment (uschi2000) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java]: done

File: atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java (hunk 4 above)
Comment (uschi2000) [inline atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java]: done

File: docs/source/release_notes/release-notes.rst (hunk 6 above)
Comment (uschi2000) [inline docs/source/release_notes/release-notes.rst]: reverted

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/Schema.java (hunk 3 above)
Comment (uschi2000) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/Schema.java]: done

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java (hunk 2 above)
Comment (uschi2000) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java]: reverted

File: atlasdb-client/src/main/java/com/palantir/atlasdb/schema/SweepSchema.java (hunk 1 above)
Comment (uschi2000) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/schema/SweepSchema.java]: done

**Comment by uschi2000** [comment]
All addressed.

**Comment by uschi2000** [comment]
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions.

---

*[atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/render/TableRendererTest.java, line 40 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVh0AYzXbyE4UmfYqjr-r1-40:-KVkER83JLVl5ed-Kni5:b-896fix) ([raw file](https://github.com/palantir/atlasdb/blob/6ab5c9c4f5999c95d66ba3c8124cb4ce069c09d0/atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/render/TableRendererTest.java#L40)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Again, this is actually two tests.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1162)*

File: atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java [hunk 9]
```diff
@@ -119,10 +120,12 @@
 public class TableRenderer {
     private final String packageName;
     private final Namespace namespace;
+    private final OptionalType optionalType;
 
-    public TableRenderer(String packageName, Namespace namespace) {
+    public TableRenderer(String packageName, Namespace namespace, OptionalType optionalType) {
         this.packageName = Preconditions.checkNotNull(packageName);
         this.namespace = Preconditions.checkNotNull(namespace);
+        this.optionalType = optionalType;
```
Comment (schlosna) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java]: nit: checkNotNull

**Comment by uschi2000** [comment]
Done

**Comment by uschi2000** [comment]
Don't know what the deal is with those timed out Circle builds?

**Comment by uschi2000** [comment]
@gsheasby ready for you.

**Comment by gsheasby** [comment]
Reviewed 5 of 15 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/OptionalType.java, line 22 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyQL_sOYaSgCmWvK7f:-KVyQL_sOYaSgCmWvK7g:b6nuv6h) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/OptionalType.java#L22)):*
> ```Java
>  */
> public enum OptionalType {
>     GUAVA("absent"),
> ```

How does Checkstyle feel about this?

---

*[atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java, line 1065 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyQqOAFqKyvGQKFbsQ:-KVyQqOBco_XF4-oZwu-:bnplkbs) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java#L1065)):*
> ```Java
> 
>         private String getOptionalTypeForClass() {
>             return "Optional<" + RowResult + ">";
> ```

nit: this can probably be inlined now. I wouldn't be against inlining the below method, too, since you always know what `value` will be and no paths are used more than once. Up to you, though.

---

*[atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java, line 1329 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyRhnae-Xbtb70fNKR:-KVyRhnae-Xbtb70fNKS:b-9jyeio) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java#L1329)):*
> ```Java
>     }
> 
>     private static final Class<?>[] IMPORTS_SANS_OPTIONAL = {
> ```

Does using the (admittedly well-known) French word add anything that "WITHOUT" doesn't?

---

*[atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java, line 47 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyT1pS47Anbvvawm5-:-KVyT1pS47Anbvvawm50:bvw3a3t) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java#L47)):*
> ```Java
>         schema.addTableDefinition("TableName", getSimpleTableDefinition(TABLE_REF));
>         schema.renderTables(testFolder.getRoot());
>         assertThat(readFileIntoString(testFolder.getRoot(), "package/TestTableTable.java"),
> ```

Magic string `"package/TestTableTable.java"` is repeated three times (and depends on `"package"`). Ditto with the three strings earlier in the method. Consider refactoring.

---

*[docs/source/release_notes/release-notes.rst, line 45 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVh2C6zz_hcc4Anegju-r1-45:-KVyU2DrEWMDeHTPHOhT:b-57h8cq) ([raw file](https://github.com/palantir/atlasdb/blob/6ab5c9c4f5999c95d66ba3c8124cb4ce069c09d0/docs/source/release_notes/release-notes.rst#L45)):*
<details><summary><i>Previously, uschi2000 (Robert Fink) wrote…</i></summary>
> reverted
</details>
LGTM. We'll have to take care which release this note actually ends up in, as we'll probably release before this change gets in.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1162)*

**Comment by uschi2000** [comment]
rebased

---

Review status: 3 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/OptionalType.java, line 22 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyQL_sOYaSgCmWvK7f:-KVzUe_hJw3EV2tkakil:b-2dhryn) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/OptionalType.java#L22)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> How does Checkstyle feel about this?
</details>
The build is passing so I assume checkstyle is fine with it? Runnig atlasdb-client:checkstyleMain locally is also green.

---

*[atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java, line 128 at r2](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVh0AYzXbyE4UmfYqjo-r2-128:-KVzV1_2QtN30OFA8doC:b-896fix) ([raw file](https://github.com/palantir/atlasdb/blob/e0f3804214cd946b818975bddf702dad36ea5b36/atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java#L128)):*
<details><summary><i>Previously, schlosna (David Schlosnagle) wrote…</i></summary>
> nit: checkNotNull
</details>
Done.

---

*[atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java, line 1065 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyQqOAFqKyvGQKFbsQ:-KVzVItZJ4cDggCp84pd:b-896fix) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java#L1065)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> nit: this can probably be inlined now. I wouldn't be against inlining the below method, too, since you always know what `value` will be and no paths are used more than once. Up to you, though.
</details>
Done.

---

*[atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java, line 1329 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyRhnae-Xbtb70fNKR:-KVzVJooHvHMXyM1LiOS:b-ufdkci) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java#L1329)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Does using the (admittedly well-known) French word add anything that "WITHOUT" doesn't?
</details>
it adds international flair and removes 3 characters. changed to WITHOUT

---

*[atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java, line 47 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyT1pS47Anbvvawm5-:-KVzWE1teS-W7NwKf0zS:b-ljdf1p) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java#L47)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Magic string `"package/TestTableTable.java"` is repeated three times (and depends on `"package"`). Ditto with the three strings earlier in the method. Consider refactoring.
</details>
moved the strings that carry function to CONSTANTS

---

*[docs/source/release_notes/release-notes.rst, line 45 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVh2C6zz_hcc4Anegju-r1-45:-KVzWLIZuQGFFZueQD_r:b-q3mpi8) ([raw file](https://github.com/palantir/atlasdb/blob/6ab5c9c4f5999c95d66ba3c8124cb4ce069c09d0/docs/source/release_notes/release-notes.rst#L45)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> LGTM. We'll have to take care which release this note actually ends up in, as we'll probably release before this change gets in.
</details>
When do you plan to release? What's the reason for not including this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-:-KVzWar4COKgkzMZgp_c:b-hm0ll1)*

**Comment by gsheasby** [comment]
Reviewed 1 of 1 files at r4, 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/OptionalType.java, line 22 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyQL_sOYaSgCmWvK7f:-KW2T41TQHuVc-BpkWKg:b-u7qutm) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/OptionalType.java#L22)):*
<details><summary><i>Previously, uschi2000 (Robert Fink) wrote…</i></summary>
> The build is passing so I assume checkstyle is fine with it? Runnig atlasdb-client:checkstyleMain locally is also green.
</details>
Cool. Checkstyle isn't actually enabled for `atlasdb-client` yet, so running checkstyleMain does nothing. However, your IDE might complain depending on its settings.

---

*[atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java, line 1329 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyRhnae-Xbtb70fNKR:-KW2UMGWUkz6JyLXQFiu:bb6o2k9) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/main/java/com/palantir/atlasdb/table/description/render/TableRenderer.java#L1329)):*
<details><summary><i>Previously, uschi2000 (Robert Fink) wrote…</i></summary>
> it adds international flair and removes 3 characters. changed to WITHOUT
</details>
:fr:

---

*[atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java, line 47 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyT1pS47Anbvvawm5-:-KW2UXCXkDEBmZtud8S9:b-d9bl4f) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java#L47)):*
<details><summary><i>Previously, uschi2000 (Robert Fink) wrote…</i></summary>
> moved the strings that carry function to CONSTANTS
</details>
Half way there... would like to see something like:
`private static final String TEST_PATH = TEST_PACKAGE + "/" + TEST_TABLE_NAME + "Table.java"`.

---

*[docs/source/release_notes/release-notes.rst, line 45 at r1](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVh2C6zz_hcc4Anegju-r1-45:-KW2Vh2aN3RNNyEhyTyB:b-8stztn) ([raw file](https://github.com/palantir/atlasdb/blob/6ab5c9c4f5999c95d66ba3c8124cb4ce069c09d0/docs/source/release_notes/release-notes.rst#L45)):*
<details><summary><i>Previously, uschi2000 (Robert Fink) wrote…</i></summary>
> When do you plan to release? What's the reason for not including this?
</details>
We cut a beta of 0.23 at the tail end of last week, for testing with an internal product. We're iterating with cutting betas and testing the internal product, and only necessary fixes should go into that branch.

Open PRs targeted at develop will go into 0.24 (or a later release if it doesn't make it).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1162)*

**Comment by gsheasby** [comment]
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java, line 47 at r3](https://reviewable.io:443/reviews/palantir/atlasdb/1162#-KVyT1pS47Anbvvawm5-:-KW2ZNhhprV9uYRCcH4u:b6khehv) ([raw file](https://github.com/palantir/atlasdb/blob/6d07741aa03d812d79eeec801b5d64f49c57466d/atlasdb-client/src/test/java/com/palantir/atlasdb/table/description/SchemaTest.java#L47)):*
<details><summary><i>Previously, gsheasby (Glenn Sheasby) wrote…</i></summary>
> Half way there... would like to see something like:
> `private static final String TEST_PATH = TEST_PACKAGE + "/" + TEST_TABLE_NAME + "Table.java"`.
</details>
Did it myself to speed up iteration due to the time difference.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/palantir/atlasdb/1162)*

**Comment by uschi2000** [comment]
Cool, I consider this done from my end and assume you are going to merge
this into whatever fits your release process. There is no way to get this
into 0.23?

**Comment by gsheasby** [comment]
No; we already have a _ton_ of stuff going in there. However, we do plan on releasing 0.24.0 this week, and that will include this change (and a bugfix for migration clis, #1171).

**Comment by uschi2000** [comment]
Great, thanks.

**Comment by uschi2000** [comment]
Ping. Anything left here?

**Pull request merged**